packager/packager.cc

```diff
@@ -0,0 +1,665 @@
+                    fmt::format("Replace `{}` with `{}`", constLit->cnst.data(ctx)->shortName(ctx), replacement),
+                    {core::AutocorrectSuggestion::Edit{core::Loc(ctx.file, nameLoc), replacement}}});
+            }
+        }
+        pName.fullName.emplace_back(constLit->cnst);
+        constLit = ast::cast_tree<ast::UnresolvedConstantLit>(constLit->scope);
+    }
+    reverse(pName.fullName.begin(), pName.fullName.end());
+    if (!pName.fullName.empty()) {
+        // Foo::Bar => Foo_Bar_Package
+        auto mangledName = absl::StrCat(absl::StrJoin(pName.fullName, "_", NameFormatter(ctx)), "_Package");
+        pName.mangledName = ctx.state.enterNameConstant(mangledName);
+    } else {
+        // Not a valid name.
+        pName.mangledName = core::NameRef::noName();
+    }
+    return pName;
+}
+
+bool isReferenceToPackageSpec(core::Context ctx, ast::TreePtr &expr) {
+    auto constLit = ast::cast_tree<ast::UnresolvedConstantLit>(expr);
+    return constLit != nullptr && constLit->cnst == core::Names::Constants::PackageSpec();
+}
+
+unique_ptr<ast::UnresolvedConstantLit> copyConstantLit(ast::UnresolvedConstantLit *lit) {
+    auto copy = lit->deepCopy();
+    // Cast from Expression to UnresolvedConstantLit.
+    auto copyUcl = ast::cast_tree<ast::UnresolvedConstantLit>(copy);
+    ENFORCE(copyUcl != nullptr);
+    unique_ptr<ast::UnresolvedConstantLit> rv(copyUcl);
+    copy.release();
+    return rv;
+}
+
+ast::TreePtr name2Expr(core::NameRef name, ast::TreePtr scope = ast::MK::EmptyTree()) {
+    return ast::MK::UnresolvedConstant(core::LocOffsets::none(), move(scope), name);
+}
+
+ast::TreePtr PackageName::getFullNameLiteral(core::LocOffsets loc) const {
+    ast::TreePtr name = ast::MK::EmptyTree();
+    for (auto namePart : fullName) {
+        name = name2Expr(namePart, move(name));
+    }
+    // Outer name should have the provided loc.
+    if (auto lit = ast::cast_tree<ast::UnresolvedConstantLit>(name)) {
+        name = ast::MK::UnresolvedConstant(loc, move(lit->scope), lit->cnst);
+    }
+    return name;
+}
+
+ast::UnresolvedConstantLit *verifyConstant(core::MutableContext ctx, core::NameRef fun, ast::TreePtr &expr) {
+    auto target = ast::cast_tree<ast::UnresolvedConstantLit>(expr);
+    if (target == nullptr) {
+        if (auto e = ctx.beginError(expr->loc, core::errors::Packager::InvalidImportOrExport)) {
+            e.setHeader("Argument to `{}` must be a constant", fun.show(ctx));
+        }
+    }
+    return target;
+}
+
+core::LocOffsets firstLineOfFile(core::Context ctx) {
+    auto file = ctx.file;
+    core::LocOffsets firstLine{0, 0};
+    // Line 0 is reserved.
+    if (file.data(ctx).lineCount() > 1) {
+        firstLine.endLoc = static_cast<u4>(file.data(ctx).lineBreaks()[1]);
+    } else {
+        firstLine.endLoc = static_cast<u4>(file.data(ctx).source().length());
+    }
+    return firstLine;
+}
+
+struct PackageInfoFinder {
+    shared_ptr<PackageInfo> info = nullptr;
```

Comment:
Could this be a `unique_ptr`?

Reply:
Yup, I believe so.

Reply:
👍 let's make it a unique_ptr then. I like to use `unique_ptr` by default, especially when there's only one thing that owns something at a given point in time.

Reply:
done